game-app/game-core/src/main/java/games/strategy/engine/delegate/IDelegateBridge.java

```diff
@@ -97,4 +100,8 @@ public interface IDelegateBridge {
   void sendMessage(WebSocketMessage webSocketMessage);
 
   ResourceLoader getResourceLoader();
+
+  default IntegerMap<UnitType> getCostsForTuv(final GamePlayer player) {
+    return new TuvCostsCalculator().getCostsForTuv(player);
+  }
```

Comment:
Hmm this seems oddly specific tbh. `getData()`, `getHistoryWriter()` all seem so generic. Also is there a reason why a new instance of `TuvCostsCalculator` is created every call to this except for the fact that interfaces can't have any fields?
If this method was changed to something like `getTuvCostsCalculator()` this would allow some callers to store the object in a local variable  and use it a couple of times so the behaviour of `DummyDelegateBridge` and the other implementations would be more consistent. What do you think?

Reply:
I agree that re-using is better, but the place where this matters is the BattleCalculator case, where we use a IDelegateBridge that overrides this version. Ideally, the other implementations ones can override it too, but those don't really matter for performance reasons. I can add a comment.

Exposing a `getTuvCostsCalculator()` will just make the call sites more verbose, but we can certainly change it if you feel strongly about that.

Reply:
One more follow up cost, I disagree with the following:

> If this method was changed to something like getTuvCostsCalculator() this would allow some callers to store the object in a local variable and use it a couple of times

Having callers having to cache the object just adds a lot of complexity, whereas the ideal implementation is the delegate bridge is what's caching the object. It's just that in practice, it only matters for the BattleCalculator case since that does something like 100k battles for every 20 battles in the real game.

Reply:
Sorry and one more bit of context - the reason we can't just trivially make it a member of e.g. DefaultDelegateBridge, is that the information could potentially change and we would need to clear the cache at appropriate times. That's a lot of extra complexity that's not needed to fix the main performance problem.

Reply:
Thanks for the context, this definitely helps understanding the issue. So the performance improvements won't work in general in a real "scenario" only for simulation, correct?

Regardless I just had an idea, not sure how feasible it is though: Did you consider moving the tuv calculating code into `GamePlayer`? Could that help with the problems you just mentioned or is this something completely different?

Reply:
I think if we want to persist on something like Gameplay or GameData, then we need to make sure to reset the cache when certain state changes which is a lot of complexity.